packages/devtools_app/lib/src/utils.dart

```diff
@@ -199,6 +199,21 @@ typedef VoidFunction = void Function();
 /// future.
 typedef VoidAsyncFunction = Future<void> Function();
 
+class SingleRunCallback {
```

Comment:
[optional] How about

```dart
void Function() runOnce(void Function() f) {
  var hasRun = false;
  return () {
    if (hasRun) return;
    hasRun = true;
    f();
  };
}
```

Then in the constructor for `InspectorController`:

```dart
  InspectorController({
      ...,
      void Function() onExpandCollapseSupported,
  }) : onExpandCollapseSupported = runOnce(onExpandCollapseSupported);
```

That way the usage doesn't need to concern itself with the wrapping, and callbacks are kept as callbacks instead of a different class.

Reply:
wrapping the inspector controller calls with runOnce seems like an easy way to get weird bugs in the future. I'd recommend instead switching these to use ValueNotifier and handling value changes robustly. For example, what if a HotReload triggers a valid change in whether onExpandCollapseSupported is true. In that case, you will get odd behavior due to the runOnce.

Reply:
This would require a refactor of `ServiceConnectionManager` to switch from streams to value notifiers, which I'm a little hesitant to do right before release and Flutter Interact. I agree this should be done though - are you okay with adding a TODO to remove the runOnce code when we refactor `ServiceConnectionManager`? 

The only way whether onExpandCollapseSupported will change is if a user switched to an older version of Flutter mid-run (which I'm not even sure would affect the running app since it was already ran with a newer version of the framework).

Reply:
![lgtm](https://i.imgur.com/JgNRalI.png)
WIth a bug and a TODO. Btw you can see a similar spot where I added a shim to also support ValueNotifier in the VMServiceManager code because I also wanted to avoid this sort of bug.

Reply:
Actually...the implementation was more straight forward than I anticipated. Switched to using ValueNotifiers. PTAL.